scripts/04-effect-for-beginners.md

```diff
@@ -0,0 +1,1622 @@
+
+---
+
+The pipe method:
+
+```ts
+pipe(effect, func1, func2, ...)
+```
+
+```ts
+effect.pipe(func1, func2, ...)
+```
+
+These are identical
+
+notes:
+
+For convenience the Effect type also has a pipe method which is identical to passing the effect as the first argument to the pipe function, but looks a bit cleaner. I'll use these interchangeable through the examples in this video, just know they are the same and you can use whichever way works best for you.
+
+---
+
+Transforming the value of an effect with map
+```ts
+// type: Effect<never, never, string>
+const mappedEffect = pipe(
+  Effect.succeed(1),
+  Effect.map((x: number) => String(x))
+)
+ 
+console.log(Effect.runSync(mappedEffect)) // Output: "1"
+```
+
+notes:
+the map function is simple, it applies the function passed to it to the value inside an effect, and returns a new effect with the value as the output of that function.
+
+In this example the function passed to map takes in a number, in this case the 1 value from the previous effect, and converts it to a string. When we log the result after running the effect and see the output is a string 1
+
+---
+```ts
+const mappedEffect = pipe(
+  Effect.succeed({ x: 5, y: 0 }),
+  Effect.map(({x, y}: { x: number, y: number }) =>
+	y === 0 ? Effect.fail(new Error()) : Effect.succeed(x/y)
+  )
+)
+```
+
+notes:
+Lets recreate our divide function from earlier, but this time with the input as a single object so it can be the singular value of an effect. Just like earlier our mapping function will either error with an Error or succeed with a number.
+
+---
+
+This isn't right
+
+```ts
+const mappedEffect = pipe(
+  Effect.succeed({ x: 5, y: 0 }),
+  Effect.map(({x, y}: { x: number, y: number }) =>
+	y === 0 ? Effect.fail(new Error()) : Effect.succeed(x/y)
+  )
+)
+```
+
+```ts
+type mappedEffect = Effect<
+	never, 
+	never, 
+	Effect<never, Error, number>
+>;
+```
+
+notes:
+Well this sure isn't what we wanted. Our Effect has a value of another Effect. 
+
+---
+We want to the Effects to merge
+```ts
+type WhatWeWant = Effect<never, Error, number>;
+```
+notes:
+What we really want it for the effects to merge, carrying the new value type from the output of the mapping function, while combining the possible Errors
+
+---
+
+Effectful transformations with flatMap
+
+```ts
+// type: Effect<never, Error, number>
+const flatMappedEffect = pipe(
+  Effect.succeed({ x: 5, y: 0 }),
+  Effect.flatMap(({x, y}: { x: number, y: number }) =>
+	y === 0 ? Effect.fail(new Error()) : Effect.succeed(x/y)
+  )
+)
+```
+
+notes:
+To do this we can use flatMap. flatMap takes a function that takes an input of the type of the value of the previous Effect and returns a new Effect. The result of the flat map will be a new Effect whose value is the value of the new Effect, and whos error and requirements are a combination of the old and new Effects.
+
+---
+
+```ts
+// returns: Effect<never, never, number>
+const getNumber = () => Effect.succeed(Math.random() * 10);
```

Comment:
this function performs a side-effect eagerly, namely calling `getNumber()` will return an effect where the random number is always the same, example: `const a = getNumber(); const b = Effect.all([a, a])` here `b` will always have first and second element equal.

It should be wrote like:

```ts
const getNumber = () => Effect.sync(() => Math.random() * 10)
```

and given as we said before that Effect is only a description this function takes no input, ergo it can be a constant:

```ts
const getNumber = Effect.sync(() => Math.random() * 10)
```